ports/esp32/machine_pwm.c

```diff
@@ -239,14 +303,30 @@ STATIC mp_obj_t esp32_pwm_deinit(mp_obj_t self_in) {
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(esp32_pwm_deinit_obj, esp32_pwm_deinit);
 
 STATIC mp_obj_t esp32_pwm_freq(size_t n_args, const mp_obj_t *args) {
+    esp32_pwm_obj_t *self = MP_OBJ_TO_PTR(args[0]);
     if (n_args == 1) {
         // get
-        return MP_OBJ_NEW_SMALL_INT(timer_cfg.freq_hz);
+        return MP_OBJ_NEW_SMALL_INT(timers[chan_timer[self->channel]].freq_hz);
     }
 
     // set
     int tval = mp_obj_get_int(args[1]);
-    if (!set_freq(tval)) {
+    cleanup_timer(self->channel, chan_timer[self->channel]);
```

Comment:
I think it's important to make it so that there is a smooth transition to another frequency, if the same timer that's currently connected to the pin/channel can be reused.  Eg, when freq changes, if the timer on this pin is not used by any other PWM channel, and there is no other timer with the requested freq, then just call `set_freq()` on the current timer.